File: src/sources.ts
import { createHash } from 'node:crypto';
import type { SearchResponse } from '@mendable/firecrawl-js';

import type {
  SourceDocument,
  SourceQuality,
  SourceType,
} from './types/research';

type SearchItem = {
  url?: string;
  markdown?: string;
  title?: string;
  description?: string;
  metadata?: {
    title?: string;
    description?: string;
    publishedTime?: string;
    publishedAt?: string;
    sourceURL?: string;
  };
};

const TRACKING_PARAMS = new Set([
  'fbclid',
  'gclid',
  'mc_cid',
  'mc_eid',
  'ref',
  'utm_campaign',
  'utm_content',
  'utm_medium',
  'utm_source',
  'utm_term',
]);

const OFFICIAL_DOMAINS = [
  'gov',
  'gov.cn',
  'gov.uk',
  'europa.eu',
  'sec.gov',
  'who.int',
  'worldbank.org',
  'stats.gov.cn',
  'caam.org.cn',
];

const RESEARCH_DOMAINS = [
  'arxiv.org',
  'doi.org',
  'nature.com',
  'sciencedirect.com',
];

const MEDIA_DOMAINS = [
  'reuters.com',
  'ft.com',
  'bloomberg.com',
  'nytimes.com',
  'wsj.com',
  'theguardian.com',
];

function shortHash(value: string) {
  return createHash('sha256').update(value).digest('hex').slice(0, 12);
}

export function canonicalizeUrl(rawUrl: string): string {
  try {
    const url = new URL(rawUrl);
    url.hash = '';
    for (const key of [...url.searchParams.keys()]) {
      if (TRACKING_PARAMS.has(key.toLowerCase())) {
        url.searchParams.delete(key);
      }
    }
    url.pathname = url.pathname.replace(/\/+$/, '') || '/';
    return url.toString();
  } catch {
    return rawUrl.trim();
  }
}

function matchesDomain(hostname: string, domain: string) {
  return hostname === domain || hostname.endsWith(`.${domain}`);
}

export function classifySource(url: string): SourceType {
  try {
    const hostname = new URL(url).hostname.toLowerCase();
    if (OFFICIAL_DOMAINS.some(domain => matchesDomain(hostname, domain))) {
      return hostname.includes('gov') || hostname.includes('sec.gov')
        ? 'regulator'
        : 'official';
    }
    if (RESEARCH_DOMAINS.some(domain => matchesDomain(hostname, domain))) {
      return 'research';
    }
    if (MEDIA_DOMAINS.some(domain => matchesDomain(hostname, domain))) {
      return 'media';
    }
    if (hostname.endsWith('.edu') || hostname.endsWith('.edu.cn')) {
      return 'research';
    }
    return 'unknown';
  } catch {
    return 'unknown';
  }
}

function scoreSource({
  url,
  title,
  content,
  publishedAt,
  sourceType,
  maxAgeDays = 730,
}: {
  url: string;
  title?: string;
  content: string;
  publishedAt?: string;
  sourceType: SourceType;
  maxAgeDays?: number;
}): SourceQuality {
  const reasons: string[] = [];
  const authority =
    sourceType === 'regulator'
      ? 1
      : sourceType === 'official' || sourceType === 'research'
        ? 0.85
        : sourceType === 'media'
          ? 0.7
          : 0.4;
  const primaryness =
    sourceType === 'regulator' || sourceType === 'official'
      ? 1
      : sourceType === 'research'
        ? 0.85
        : 0.55;
  const specificity = Math.min(
    1,
    (content.length / 3000) * 0.6 + (title ? 0.4 : 0),
  );

  let freshness = 0.35;
  if (publishedAt) {
    const published = Date.parse(publishedAt);
    if (!Number.isNaN(published)) {
      const ageDays = Math.max(0, (Date.now() - published) / 86_400_000);
      freshness = Math.max(0, 1 - ageDays / maxAgeDays);
      reasons.push(`published ${Math.round(ageDays)} days ago`);
    }
  } else {
    reasons.push('publication date unavailable');
  }

  if (authority >= 0.85) reasons.push(`source type: ${sourceType}`);
  if (specificity >= 0.7) reasons.push('contains substantial source content');
  const score = Math.round(
    (authority * 0.3 +
      primaryness * 0.25 +
      freshness * 0.2 +
      specificity * 0.25) *
      100,
  );

  return { authority, primaryness, freshness, specificity, score, reasons };
}

export function sourcesFromSearchResponse(
  result: SearchResponse,
  maxAgeDays?: number,
): SourceDocument[] {
  const fetchedAt = new Date().toISOString();
  const sources = new Map<string, SourceDocument>();

  for (const rawItem of result.data as SearchItem[]) {
    if (!rawItem.url || !rawItem.markdown?.trim()) continue;
    const canonicalUrl = canonicalizeUrl(rawItem.url);
    const sourceType = classifySource(canonicalUrl);
    const title = rawItem.title ?? rawItem.metadata?.title;
    const publishedAt =
      rawItem.metadata?.publishedAt ?? rawItem.metadata?.publishedTime;
    const source: SourceDocument = {
      id: `src-${shortHash(canonicalUrl)}`,
      url: rawItem.url,
      canonicalUrl,
      title,
      publisher: rawItem.metadata?.sourceURL,
      publishedAt,
      fetchedAt,
      sourceType,
      content: rawItem.markdown.trim(),
      quality: scoreSource({
        url: canonicalUrl,
        title,
        content: rawItem.markdown,
        publishedAt,
        sourceType,
        maxAgeDays,
      }),
    };
    sources.set(canonicalUrl, source);
  }

  return [...sources.values()];
}

File: evidence_research/sources.py
from __future__ import annotations

import hashlib
from datetime import UTC, datetime
from urllib.parse import parse_qsl, urlencode, urlsplit, urlunsplit

from .models import SourceDocument, SourceQuality, SourceType

TRACKING_PARAMS = {
    "fbclid",
    "gclid",
    "mc_cid",
    "mc_eid",
    "ref",
    "utm_campaign",
    "utm_content",
    "utm_medium",
    "utm_source",
    "utm_term",
}

OFFICIAL_DOMAINS = {
    "gov",
    "gov.cn",
    "gov.uk",
    "europa.eu",
    "sec.gov",
    "who.int",
    "worldbank.org",
    "stats.gov.cn",
    "caam.org.cn",
}
RESEARCH_DOMAINS = {"arxiv.org", "doi.org", "nature.com", "sciencedirect.com"}
MEDIA_DOMAINS = {
    "reuters.com",
    "ft.com",
    "bloomberg.com",
    "nytimes.com",
    "wsj.com",
    "theguardian.com",
}


def _short_hash(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()[:12]


def canonicalize_url(raw_url: str) -> str:
    try:
        parts = urlsplit(raw_url.strip())
        query = [
            (key, value)
            for key, value in parse_qsl(parts.query, keep_blank_values=True)
            if key.lower() not in TRACKING_PARAMS
        ]
        path = parts.path.rstrip("/") or "/"
        return urlunsplit(
            (
                parts.scheme.lower(),
                parts.netloc.lower(),
                path,
                urlencode(query),
                "",
            )
        )
    except ValueError:
        return raw_url.strip()


def _matches_domain(hostname: str, domain: str) -> bool:
    return hostname == domain or hostname.endswith(f".{domain}")


def classify_source(url: str) -> SourceType:
    hostname = urlsplit(url).hostname or ""
    hostname = hostname.lower()
    if any(_matches_domain(hostname, domain) for domain in OFFICIAL_DOMAINS):
        return SourceType.REGULATOR if "gov" in hostname or hostname == "sec.gov" else SourceType.OFFICIAL
    if any(_matches_domain(hostname, domain) for domain in RESEARCH_DOMAINS):
        return SourceType.RESEARCH
    if any(_matches_domain(hostname, domain) for domain in MEDIA_DOMAINS):
        return SourceType.MEDIA
    if hostname.endswith((".edu", ".edu.cn")):
        return SourceType.RESEARCH
    return SourceType.UNKNOWN


def _parse_date(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
        return parsed if parsed.tzinfo else parsed.replace(tzinfo=UTC)
    except ValueError:
        return None


def score_source(
    *,
    url: str,
    title: str | None,
    content: str,
    published_at: str | None,
    source_type: SourceType,
    max_age_days: int = 730,
) -> SourceQuality:
    reasons: list[str] = []
    authority = {
        SourceType.REGULATOR: 1.0,
        SourceType.OFFICIAL: 0.85,
        SourceType.RESEARCH: 0.85,
        SourceType.MEDIA: 0.7,
    }.get(source_type, 0.4)
    primaryness = {
        SourceType.REGULATOR: 1.0,
        SourceType.OFFICIAL: 1.0,
        SourceType.RESEARCH: 0.85,
    }.get(source_type, 0.55)
    specificity = min(1.0, len(content) / 3000 * 0.6 + (0.4 if title else 0))

    freshness = 0.35
    published = _parse_date(published_at)
    if published:
        age_days = max(0, (datetime.now(UTC) - published).days)
        freshness = max(0.0, 1 - age_days / max_age_days)
        reasons.append(f"published {age_days} days ago")
    else:
        reasons.append("publication date unavailable")

    if authority >= 0.85:
        reasons.append(f"source type: {source_type.value}")
    if specificity >= 0.7:
        reasons.append("contains substantial source content")

    score = round((authority * 0.3 + primaryness * 0.25 + freshness * 0.2 + specificity * 0.25) * 100)
    return SourceQuality(authority, primaryness, freshness, specificity, score, reasons)


def source_from_item(item: dict[str, object], max_age_days: int = 730) -> SourceDocument | None:
    raw_url = item.get("url")
    content = item.get("markdown") or item.get("content")
    if not isinstance(raw_url, str) or not isinstance(content, str) or not content.strip():
        return None

    metadata = item.get("metadata")
    metadata = metadata if isinstance(metadata, dict) else {}
    title = item.get("title") or metadata.get("title")
    published_at = metadata.get("publishedAt") or metadata.get("publishedTime")
    canonical_url = canonicalize_url(raw_url)
    source_type = classify_source(canonical_url)
    title = title if isinstance(title, str) else None
    published_at = published_at if isinstance(published_at, str) else None
    return SourceDocument(
        id=f"src-{_short_hash(canonical_url)}",
        url=raw_url,
        canonical_url=canonical_url,
        content=content.strip(),
        source_type=source_type,
        fetched_at=datetime.now(UTC).isoformat(),
        title=title,
        publisher=metadata.get("sourceURL") if isinstance(metadata.get("sourceURL"), str) else None,
        published_at=published_at,
        quality=score_source(
            url=canonical_url,
            title=title,
            content=content,
            published_at=published_at,
            source_type=source_type,
            max_age_days=max_age_days,
        ),
    )


def sources_from_search(items: list[dict[str, object]], max_age_days: int = 730) -> list[SourceDocument]:
    sources: dict[str, SourceDocument] = {}
    for item in items:
        source = source_from_item(item, max_age_days=max_age_days)
        if source:
            sources[source.canonical_url] = source
    return list(sources.values())
